File: flutter-project/lib/api/request.dart
import 'dart:convert';
import 'package:dio/dio.dart';

//请求地址
const baseUrl = "http://localhost";

class Request {
  static BaseOptions options = BaseOptions(
    baseUrl: baseUrl,
    connectTimeout: 15000, //请求时间
    receiveTimeout: 10000, //响应时间
  );
  static Dio dio = Dio(options);

  static Future request<T>(String url, String method,
      {Map? params, data}) async {
    if (params != null) {
      url = url + '?';
      params.forEach((key, value) {
        url = url + '$key=' + '$value&';
      });
      url = url.substring(0, url.length - 1);
    }

    try {
      Response res =
          await dio.request(url, data: data, options: Options(method: method));
      if (res.statusCode == 200 || res.statusCode == 201) {
        try {
          if (res.data['status'] != 200) {
            return Future.error(res.data['msg']);
          } else {
            if (res.data is Map) {
              return res.data;
            } else {
              return json.decode(res.data);
            }
          }
        } catch (e) {
          return Future.error('解析响应数据异常');
        }
      } else {
        return Future.error('HTTP错误');
      }
    } catch (e) {
      return Future.error('服务器无法连接');
    }
  }

  //封装GET请求
  static Future get<T>(String url, {Map? params}) {
    return request(url, 'get', params: params);
  }

  //封装POST请求
  static Future post<T>(String url, {Map? params, data}) {
    return request(url, 'post', params: params, data: data);
  }
}

File: vite-vue3/src/api/request.js
import axios from 'axios';
import router from '../router/index.js';

const instance = axios.create({
    baseURL: import.meta.env.VITE_API_URL,
    timeout: 15000, withCredentials: true,
    headers: {post: {'Content-Type': 'application/json;charset=utf-8'}}
});
// 请求拦截器
instance.interceptors.request.use(
    config => {
        const accessToken = localStorage.getItem('accessToken');
        if (!['/login', '/register'].includes(config.url) && !accessToken) {
            const controller = new AbortController();
            controller.abort();
            localStorage.removeItem('accessToken');
            router.push({path: '/login'}).then(res => console.log(res));
            return Promise.reject({code: 500});
        }
        if (accessToken) config.headers.Authorization = `Bearer ${accessToken}`;
        return config;
    },
    error => (Promise.error(error))
);
// 响应拦截器
instance.interceptors.response.use(
    response => {
        if (response.status === 200) return Promise.resolve(response);
        else return Promise.reject(response);
    },
    // 服务器状态码不是200的情况
    async error => {
        const originalRequest = error.config;
        if (error.status === 401 && !originalRequest._retry) {
            originalRequest._retry = true;
            const {data} = await instance.post('/refreshToken', {});
            localStorage.setItem('accessToken', data.accessToken || null);
            originalRequest.headers.Authorization = `Bearer ${data.accessToken}`;
            return instance(originalRequest);
        }
        if (error.status === 403) {
            localStorage.removeItem('accessToken');
            await router.push({path: '/login'});
        }
        return Promise.reject(error.response);
    }
);
const pending = new Set();

export function get(url, params) {
    if (pending.has(url)) return Promise.resolve({});
    pending.add(url);
    return new Promise((resolve, reject) => {
        instance.get(url, {params}).then(res => {
            pending.delete(url);
            resolve(res.data);
        }).catch(err => {
            pending.delete(url);
            reject(err);
        });
    });
}

export function post(url, params) {
    return new Promise((resolve, reject) => {
        instance.post(url, params).then(res => resolve(res.data)).catch(err => reject(err));
    });
}

export default instance;
